### PR TITLE
kilosort2_new sorter wrapper - WIP

### PR DESCRIPTION
**Pull request by magland**
This is completely untested. Just a start.

Removed dependency on matlab-npy (use plain text file output for now)
Cleaned up code, use static type declarations
Simplified install check

### CODE REVIEW

File: spikesorters/kilosort2/kilosort2_new.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        source_dir = Path(Path(__file__).parent)
+
+        p = self.params
+
+        if not check_if_installed(Kilosort2SorterNew.kilosort2_path):
+            raise Exception(Kilosort2SorterNew.installation_mesg)
+        assert isinstance(Kilosort2SorterNew.kilosort2_path, str)
+
+        # prepare electrode positions
+        electrode_dimensions = p['electrode_dimensions']
```
Comment (alejoe91) [inline spikesorters/kilosort2/kilosort2_new.py]: I think that in general the 3d geometries should be dealt with outside the specific sorters? 
What do you think?

That was done mainly for MEArec, bu now the extractor itself dumps the 3rd dimension

File: spikesorters/kilosort2/kilosort2_new.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        if electrode_dimensions is None:
+            electrode_dimensions = [0, 1]
+        if 'group' in recording.get_channel_property_names():
+            groups = [recording.get_channel_property(
```
Comment (alejoe91) [inline spikesorters/kilosort2/kilosort2_new.py]: you can also use:

`recording.get_channel_groups()`

File: spikesorters/kilosort2/kilosort2_new.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+                ch, 'group') for ch in recording.get_channel_ids()]
+        else:
+            groups = [1] * recording.get_num_channels()
+        if 'location' in recording.get_channel_property_names():
+            positions = np.array([recording.get_channel_property(
```
Comment (alejoe91) [inline spikesorters/kilosort2/kilosort2_new.py]: same here:
`positions = np.array(recording.get_channel_positions())`

File: spikesorters/kilosort2/kilosort2_new.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        # read the template txt files
+        with (source_dir / 'kilosort2_master_new.m').open('r') as f:
+            kilosort2_master_txt = f.read()
+        with (source_dir / 'kilosort2_config_new.m').open('r') as f:
+            kilosort2_config_txt = f.read()
+        with (source_dir / 'kilosort2_channelmap_new.m').open('r') as f:
+            kilosort2_channelmap_txt = f.read()
+
+        # make substitutions in txt files
+        kilosort2_master_txt = kilosort2_master_txt.format(
```
Comment (alejoe91) [inline spikesorters/kilosort2/kilosort2_new.py]: nice :)

File: spikesorters/kilosort2/kilosort2_master_new.m [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    % final splits by amplitudes
+    rez = splitAllClusters(rez, 0);
+
+    % decide on cutoff
+    rez = set_cutoff(rez);
+
+    fprintf('found %d good units \n', sum(rez.good>0))
+
+    % write the output
+    timestamps = rez.st3(:,1); %time
```
Comment (alejoe91) [inline spikesorters/kilosort2/kilosort2_master_new.m]: I see your point here, but also @samuelgarcia has a point.

Maybe we should keep the spike sorter specific formats?

**Comment by samuelgarcia** [comment]
The 
```
catch
    fprintf('----------------------------------------');
    fprintf(lasterr());
    quit(1);
```
is really cool because we can now errors!! Yeah.

**Comment by samuelgarcia** [comment]
The idea of having natural storage for a sorter is important for me.

A user could use tis code only for batch processing files with KS2. What he would except is to get folder as "before". 

I have to admit that I don't known what is the offical format for KS2, but if there is one it should be kept so a user could open the open the folder and check immediatly in its own pipline if it is OK.
Forcing our txt format for output could be problematic I think.

I totally understand the point of the NPZ matlab code dependency which is annoying.

**Comment by samuelgarcia** [comment]
@alejoe91  : Why do we have still so much travis in parralel for on PR ?

**Comment by alejoe91** [comment]
I trigger when a new PR is made and when it's pushed. You can change this in settings -> webhooks

**Comment by alejoe91** [comment]
@magland kilosort2_new does not find the kilosort2 path

**Comment by magland** [comment]
I agree with you guys. So the question is... what is the natural output format for ks2? And can we get away with not using matlab-npy?

Another possibility: hard-code matlab-npy into our code base, it's very small.

**Comment by magland** [comment]
@alejoe91 , what do you mean it does not find kilosort2_path? Are you using KILOSORT2_PATH environment variable? Are you using set_kilosort2_path? (It is no longer a sorting parameter.)

File: spikesorters/kilosort2/kilosort2_new.py (hunk 2 above)
Comment (magland) [reply spikesorters/kilosort2/kilosort2_new.py]: ok

File: spikesorters/kilosort2/kilosort2_new.py (hunk 3 above)
Comment (magland) [reply spikesorters/kilosort2/kilosort2_new.py]: ok

File: spikesorters/kilosort2/kilosort2_master_new.m (hunk 5 above)
Comment (magland) [reply spikesorters/kilosort2/kilosort2_master_new.m]: see below

File: spikesorters/kilosort2/kilosort2_new.py (hunk 1 above)
Comment (magland) [reply spikesorters/kilosort2/kilosort2_new.py]: that's up to you guys.

**Comment by mhhennig** [comment]
Aside while you're at this: I found that KS2 (or the wrapper) rounds the sampling rate down to two digits. It's annoying for comparisons! Is there a fix?

**Comment by alejoe91** [comment]
so the natural ouput it's actually the phy format at the moment. I think it's fair to hardcode npy-matlab. we only need two functions I believe ([writeNPY](https://github.com/kwikteam/npy-matlab/blob/master/npy-matlab/writeNPY.m) and [constructNPYheader](https://github.com/kwikteam/npy-matlab/blob/master/npy-matlab/constructNPYheader.m)).

@magland we could have them in a kilosort_npy_utils folder so we don't have to duplicate them for ks1 and ks2

**Comment by magland** [comment]
okay, good idea. But rather than referring to the source file, we should treat them like more template code and copy those into the output folder. (I have a reason for this that has to do with mountaintools).

**Comment by alejoe91** [comment]
I fixed it. You have to add the `installed` variable in the SorterClass (when running the run_kilosort2_new).
Can I commit my changes? It's working now on my side

**Comment by magland** [comment]
Yes, please!

**Comment by magland** [comment]
i'll be away from computer for a while. If one of you wants to change to use phy output format that would be great, otherwise I can do it later today or tomorrow.

**Comment by alejoe91** [comment]
I can do that :) ok basically doing the same as the master and config files. Did I understand correctly?

**Comment by magland** [comment]
Yes. And they will go in a directory inside the output folder

**Comment by magland** [comment]
Great. Once you merge this I can point to the new version from spikeforest and run everything -- verify that we get same or similar results.

**Comment by alejoe91** [comment]
I think it's ready to merge! Next, we should really fix ironclust (we miss it over here !)

**Comment by magland** [comment]
I agree. I wanted 'warm up' first on kilosort. I will work on ironclust later today.

**Comment by alejoe91** [comment]
One last thing: I would keep the kilosort2_path option. I think that the first time it could be passed by the user and set the environment variable internally. What do you think @magland?

**Comment by magland** [comment]
> One last thing: I would keep the kilosort2_path option. I think that the first time it could be passed by the user and set the environment variable internally. What do you think @magland?

Do you mean kilosort2_path as a sorting parameter? That's up to you. In spikeforest I do not include such things as sorting parameters, because they are more to do with the environment. I do have a mechanism to mark parameters as "environment" parameters. You may consider to something similar. Num_workers could also be considered an environment parameter, because in theory it does not affect the output.

**Comment by alejoe91** [comment]
Yep. this one: https://github.com/SpikeInterface/spikesorters/blob/matlab_wrappers/spikesorters/kilosort/kilosort.py#L41

and it's also used [here](https://github.com/SpikeInterface/spikesorters/blob/matlab_wrappers/spikesorters/kilosort/kilosort.py#L57-L66) to set an environment variable

**Comment by magland** [comment]
It should be done carefully IMO. I don't think it's a good idea to set an environment variable inside a function. Not a good idea to modify the environment non-explicitly.

The parameter, if passed, should take precedence. Then set_kilosort2_path static method. Then environment variable.

**Comment by alejoe91** [comment]
So you think the set_kilosort2_path should NOT set an environment variable? Maybe you are right :)

**Comment by magland** [comment]
> So you think the set_kilosort2_path should NOT set an environment variable?

Yes it should not. I think it should set the static kilosort2_path variable attached to the class.

**Comment by alejoe91** [comment]
got it

**Comment by magland** [comment]
But there is a danger... if the sorter is called in a subprocess, I think the static variable will not apply. Therefore, I think the recommended way is for the user to explicitly set the environment variable using os.environ['KILOSORT2_PATH']=... or something. It's easy enough to do. Maybe remove the set_kilosort2_path() static method.

**Comment by magland** [comment]
Second rule of zen of python https://www.python.org/dev/peps/pep-0020/
explicit is better than implicit

**Comment by magland** [comment]
But I'm not completely sure about it (sorry for wavering) :)

**Comment by alejoe91** [comment]
in many university computers users don't have root access, so it might be good to keep the path option as well IMO

**Comment by magland** [comment]
what does root access have to do with it?

**Comment by alejoe91** [comment]
setting environment variables. On windows to set a system variable you need admin access. You can set an environment for a specific bash/conda session, but it could be a bit of a hassle to do it everytime. Users may prefer to pass the kilosort paths

**Comment by magland** [comment]
okay

**Comment by magland** [comment]
but wouldn't os.path.environ['...'] work in windows?

**Comment by alejoe91** [comment]
I think so.. so should we have os.path.environ['...'] in the set_kilosort_path function?

**Comment by alejoe91** [comment]
I'll make a `try` statement :)

**Comment by magland** [comment]
i we are misunderstanding each other slightly. let's talk live at some
point, but for now you can do what you want.

**Comment by alejoe91** [comment]
I'm available now if you have 5 min :)

**Comment by magland** [comment]
@alejoe91 it appears that I may need to expose some more parameters for kilosort2 (the spikeforest wrappers has some additional). I'll add those.

**Comment by magland** [comment]
Also, we are currently using morley's fork of ks2, so we'll need to investigate about what is different, etc.

```
            repo='https://github.com/alexmorley/Kilosort2',
            commit='43cbbfff89b9c88cdeb147ffd4ac35bfde9c7956'
```

**Comment by samuelgarcia** [comment]
Finally you did not use the ShellScript class ?

**Comment by alejoe91** [comment]
I guess we will eventually. This was just a clean up :)

**Comment by samuelgarcia** [comment]
OK. good clean at the end.
happy that we can catch error!

**Comment by magland** [comment]
It wasn't super necessary because most of the stuff was in .m files already. But for ironclust (working on now), it will have ShellScript. (And we'll need to test in windows)
J

**Comment by magland** [comment]
I made a PR for ironclust. The unit test from spikeforest passed. Soon I will test it on all the spikeforest recordings.

**Comment by samuelgarcia** [comment]
Could you also test with unitest in spikesorters ?
For matlab sorter there are not tested on travis of course.

**Comment by magland** [comment]
unit test in spikesorters seems to work

**Comment by samuelgarcia** [comment]
Cool.

**Pull request merged**